akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala

```diff
@@ -703,7 +908,45 @@ final class ClusterReceptionist(pubSubMediator: ActorRef, settings: ClusterRecep
       }
 
     case _: MemberEvent ⇒ // not of interest
+
+    case GetClusterClients ⇒
+      sender() ! ClusterClients(clientInteractions.keySet)
+
+    case CheckDeadlines ⇒
+      clientInteractions = clientInteractions.filter {
+        case (_, failureDetector) ⇒
+          failureDetector.isAvailable
+      }
+      publishClientsUnreachable()
+  }
+
+  def updateClientInteractions(client: ActorRef): Unit =
+    clientInteractions.get(client) match {
+      case Some(failureDetector) ⇒
+        failureDetector.heartbeat()
+      case None ⇒
+        val failureDetector = new DeadlineFailureDetector(acceptableHeartbeatPause, heartbeatInterval)
+        clientInteractions = clientInteractions + (client -> failureDetector)
+    }
+
+  def publishClientsUp(): Unit = {
+    val publishableClients = clientInteractions.keySet.to[HashSet]
+    (publishableClients -- clientsPublished)
+      .foreach { c ⇒
+        log.info("Received new contact from [{}]", c.path)
```

Comment:
always difficult to pick log level, but I think it can be too verbose on server side, given that there might be 1000s of clients coming and going